README.md
<!--
Copyright 2021 Ludan Stoecklé
SPDX-License-Identifier: CC-BY-4.0
-->
# Power BI RosaeNLG plugin

This repo contains a Power BI custom visual to add RosaeNLG ability to Power BI dashboards. It is useful if you wish to generate texts using NLG (Natural Language Generation) in your PowerBI dashboards.

Current status is **Proof Of Concept**: we show that is works. You should not see the current implementation as an off-the-shelf plugin but rather as a starting point for your own plugin, which will contain your custom RosaeNLG templates. 

The Power BI RosaeNLG plugin project uses the Apache-2.0 license. It is based on the CircleCard sample code, which is licensed under the MIT license. A copy of each license can be found in the LICENSE file.

Also check:
- [Generate textual narratives in Power BI using RosaeNLG](https://medium.com/@ludan.stoeckle/generate-textual-narratives-in-power-bi-using-rosaenlg-48f81c53db39) on Medium
- [RosaeNLG main doc](https://rosaenlg.org)


## Approach

It is based on the standard Power BI tutorial where RosaeNLG has been plugged on. It actually contains little new code compared to the tutorial.

Compared to other NLG approaches, the text rendering is made locally, directly in Power BI (Desktop or online version), **without any API call**. As no data is sent to an API over the internet, RosaeNLG is a very secure NLG solution for Power BI.


## How to create a Power BI plug-in using RosaeNLG

It takes 2 major steps:
- **Create the NLG templates:** define how to produce the texts based on the PowerBI data. Do this in a separate standard RosaeNLG project - see [RosaeNLG main doc](https://rosaenlg.org).
- **Package the NLG templates in a Power BI plug-in:** it essentially consists of defining the parameters of your plug-in so that it can be plugged in a PowerBI dashboard.

If the data needs to be analysed or prepared prior to text generation, it is convenient to do it in JavaScript, either in the NLG templates or in the Power BI plug-in.


## Running the Proof Of Concept

### Plug PowerBI and RosaeNLG together

Follow the [standard Power BI tutorial](https://docs.microsoft.com/en-us/power-bi/developer/visuals/custom-visual-develop-tutorial):
- *Setting up the developer environment:* follow as described in the tutorial
- *Creating a custom visual:* instead of using the CircleCard project, checkout this repo, then `run npm install` and `pbiviz start`
- *Adding visual elements:* follow as described in the tutorial

Obtain this result:

![NLG plug-in just installed](doc/pbi_placeholder.png)

This Power BI plug-in takes 2 parameters: *sales total qty* and *sales total amount*. Let’s map them (drag and drop):

![NLG plug-in properly plugged](doc/pbi_plugged.png)

This simply means that both the Power BI plug-in and the RosaeNLG renderer are well plugged together, but this is not text generation yet.


### Render some texts

For this demo, we will use the [RosaeNLG browser IDE](https://rosaenlg.org/ide/index.html). For more complex projects, it is better to [use VSCode to author, version and compile your RosaeNLG templates](https://rosaenlg.org/rosaenlg/2.1.7/dev_experience.html#_large_projects_using_vscode).

This basic template contains 2 synonymic alternatives which essentially output the parameters in text:
```pug
- 
  // default value if not provided in the parameters
  data = data || {
    salesQty: 1000,
    salesAmount: 500000,
  }

synz
  syn
    | #[+value(data.salesQty)] items were sold
    | for a total amount of #[+value(data.salesAmount, {'FORMAT': '0a$'})]
  syn
    | total sales were #[+value(data.salesAmount, {'FORMAT': '0a$'})] for #[+value(data.salesQty)] items
| .
```
In the browser IDE, you should get this rendered text: **1,000 items were sold for a total amount of 500k$.**

Use **Export > Compiled JavaScript > for node.js** button to get the template compiled as a JavaScript function. Put the file in the powerbidemo folder and name it `compiledTemplateSales.js` (do not keep `fruits.js` or whatever).

Just refresh the dashboard and navigate in the data:

![Plug-in with real NLG text](doc/pbi_realnlg.png)

The text is still very basic, but there is not limit on the quality and complexity of the texts that you can generate!


## Inside the Power BI plug-in — how it works

In `externalJS` section of `pbiviz.json`, there is a reference to the already tinified version of the RosaeNLG renderer. Change the reference to change the language and the version. Version must be aligned with the RosaeNLG version in `package.json`.


In `visual.ts`, import both the RosaeNLG lib (`../node_modules/rosaenlg/dist/rollup/rosaenlg_tiny_en_US_XX.XX.XX.js`) and the compiled template (`../compiledTemplateSales.js`).

The `capabilities.json` file define the input parameters of your plug-in and your template. Complete that file to add parameters. Use `visual.ts` to make some mapping:

```
const data = {
 salesQty: dataView.categorical.values[0].values[0],
 salesAmount: dataView.categorical.values[1].values[0],
};
```

As an exercise you may add the sales date to the parameters and output it in the texts.

As any Power BI plug-in, it can be [packaged](https://docs.microsoft.com/en-us/power-bi/developer/visuals/custom-visual-develop-tutorial-format-options#packaging-the-custom-visual) and be run both in Desktop and Cloud versions of Power BI.



CHANGELOG.md
<!--
Copyright 2021 Ludan Stoecklé
SPDX-License-Identifier: CC-BY-4.0
-->
# Changelog

<!---
## [Unreleased]

### Added

### Fixed

### Changed

-->

## 1.1.0

### Fixed

- updated to work with recent versions of Power BI
- updated to work with a recent version of RosaeNLG


## 1.0.0

### Added

- CHANGELOG, CODE_OF_CONDUCT, COMMUNITY, CONTRIBUTING, GOVERNANCE, RELEASE, SECURITY, SUPPORT

### Changed

- README
- clear licenses between MIT and Apache-2.0



src/visual.ts
/**
 * @license
 * Copyright 2021 Ludan Stoecklé, (c) Microsoft Corporation
 * SPDX-License-Identifier: Apache-2.0 AND MIT
 */

"use strict";

import "core-js/stable";
import "./../style/visual.less";
import powerbi from "powerbi-visuals-api";
import VisualConstructorOptions = powerbi.extensibility.visual.VisualConstructorOptions;
import VisualUpdateOptions = powerbi.extensibility.visual.VisualUpdateOptions;
import IVisual = powerbi.extensibility.visual.IVisual;
import DataView = powerbi.DataView;

import rosaenlg from '../node_modules/rosaenlg/dist/rollup/rosaenlg_tiny_en_US_3.3.1.js';
import template from '../compiledTemplateSales.js';

export class Visual implements IVisual {

    private target: HTMLElement;
    private textNode: Text;

    constructor(options: VisualConstructorOptions) {

        console.log('Visual constructor', options);
        this.target = options.element;
        if (document) {
            const new_p: HTMLElement = document.createElement("div");
            this.textNode = document.createTextNode('NLG placeholder...');
            new_p.appendChild(this.textNode);
            this.target.appendChild(new_p);
        }
    }

    public update(options: VisualUpdateOptions) {
        let dataView: DataView = options.dataViews[0];
        // console.log(dataView);


        const data = {
            salesQty: dataView.categorical.values[0].values[0],
            salesAmount: dataView.categorical.values[1].values[0],
        };

        console.log(data);

        const generatedText = template({
            util: new rosaenlg.NlgLib({ language: 'en_US' }),
            data: data,
        });

        // console.log(generatedText);

        if (this.textNode) {
            this.textNode.textContent = generatedText;
        }

    }

}

src/settings.ts
/**
 * @license
 * Copyright (c) Microsoft Corporation
 * SPDX-License-Identifier: MIT
 */

// note: this is, as-is, the file provided in the CircleCard example

"use strict";

import { formattingSettings } from "powerbi-visuals-utils-formattingmodel";

import FormattingSettingsCard = formattingSettings.Card;
import FormattingSettingsSlice = formattingSettings.Slice;
import FormattingSettingsModel = formattingSettings.Model;

/**
 * Data Point Formatting Card
 */
class DataPointCardSettings extends FormattingSettingsCard {
    defaultColor = new formattingSettings.ColorPicker({
        name: "defaultColor",
        displayName: "Default color",
        value: { value: "" }
    });

    showAllDataPoints = new formattingSettings.ToggleSwitch({
        name: "showAllDataPoints",
        displayName: "Show all",
        value: true
    });

    fill = new formattingSettings.ColorPicker({
        name: "fill",
        displayName: "Fill",
        value: { value: "" }
    });

    fillRule = new formattingSettings.ColorPicker({
        name: "fillRule",
        displayName: "Color saturation",
        value: { value: "" }
    });

    fontSize = new formattingSettings.NumUpDown({
        name: "fontSize",
        displayName: "Text Size",
        value: 12
    });

    name: string = "dataPoint";
    displayName: string = "Data colors";
    slices: Array<FormattingSettingsSlice> = [this.defaultColor, this.showAllDataPoints, this.fill, this.fillRule, this.fontSize];
}

/**
* visual settings model class
*
*/
export class VisualFormattingSettingsModel extends FormattingSettingsModel {
    // Create formatting settings model formatting cards
    dataPointCard = new DataPointCardSettings();

    cards = [this.dataPointCard];
}


compiledTemplateSales.js
function pug_escape(e){var a=""+e,t=pug_match_html.exec(a);if(!t)return e;var r,c,n,s="";for(r=t.index,c=0;r<a.length;r++){switch(a.charCodeAt(r)){case 34:n="&quot;";break;case 38:n="&amp;";break;case 60:n="&lt;";break;case 62:n="&gt;";break;default:continue}c!==r&&(s+=a.substring(c,r)),c=r+1,s+=n}return c!==r?s+a.substring(c,r):s}
var pug_match_html=/["&<>]/;function template(locals) {const embeddedLinguisticResources = {"verbs":{},"words":{},"adjectives":{}};var pug_html = "", protect_stack = [], pug_mixins = {}, pug_interp;;
    var locals_for_with = (locals || {});
    
    (function (JSON, Object, console, data, embeddedLinguisticResources, params, synHelper1, util) {
      /**
 * @license
 * Copyright 2021 Ludan Stoecklé
 * SPDX-License-Identifier: Apache-2.0
 */
//-------- allows core to communicate with the inner of the template mechanic already used by Pug (i.e. pug_html)
util.setSpy({
  getPugHtml: function () {
    return pug_html;
  },
  setPugHtml: function (new_pug_html) {
    pug_html = new_pug_html;
  },
  appendPugHtml: function (append) {
    pug_html = pug_html + append;
  },
});
util.setEmbeddedLinguisticResources(embeddedLinguisticResources);
/**
 * @license
 * Copyright 2021 Ludan Stoecklé
 * SPDX-License-Identifier: Apache-2.0
 */
//-------- using GenderNumberManager
function setRefGender(obj, gender) {
  util.genderNumberManager.setRefGender(obj, gender, null);
}
function getRefGender(obj) {
  return util.genderNumberManager.getRefGender(obj);
}
function setRefNumber(obj, number) {
  util.genderNumberManager.setRefNumber(obj, number);
}
function getRefNumber(obj) {
  return util.genderNumberManager.getRefNumber(obj);
}
function setRefGenderNumber(obj, gender, number) {
  util.genderNumberManager.setRefGenderNumber(obj, gender, number);
}
function getAnonymous(gender, number) {
  return util.genderNumberManager.getAnonymous(gender, number);
}
function getAnonMS() {
  return util.genderNumberManager.getAnonMS();
}
function getAnonMP() {
  return util.genderNumberManager.getAnonMP();
}
function getAnonFS() {
  return util.genderNumberManager.getAnonFS();
}
function getAnonFP() {
  return util.genderNumberManager.getAnonFP();
}
//-------- using RefsManager
function resetRep(obj) {
  util.refsManager.resetRep(obj);
}
function hasTriggeredRef(obj) {
  return util.refsManager.hasTriggeredRef(obj);
}
function setTriggeredRef(obj) {
  util.refsManager.setTriggeredRef(obj);
}
function dumpRefMap() {
  util.refsManager.dumpRefMap();
}
function getNextRep(obj, params) {
  return util.refsManager.getNextRep(obj, params);
}
//-------- using SaidManager
function dumpHasSaid() {
  util.saidManager.dumpHasSaid();
}
function getDumpHasSaid() {
  return util.saidManager.getDumpHasSaid();
}
function getHasSaidCopy() {
  return util.saidManager.getHasSaidCopy();
}
function recordSaid(key) {
  util.saidManager.recordSaid(key);
}
function deleteSaid(key) {
  util.saidManager.deleteSaid(key);
}
function hasSaid(key) {
  return util.saidManager.hasSaid(key);
}
function recordValue(key, value) {
  util.getSaidManager().recordValue(key, value);
}
function deleteValue(key) {
  util.getSaidManager().deleteValue(key);
}
function getValue(key) {
  return util.getSaidManager().getValue(key);
}
//-------- using helper
function hasFlag(params, flag) {
  return util.helper.hasFlag(params, flag);
}
function getFlagValue(params, flag) {
  return util.helper.getFlagValue(params, flag);
}
function protectString(string) {
  return util.helper.protectString(string);
}
function isSentenceStart() {
  return util.helper.isSentenceStart();
}
function getMorF(table, obj) {
  return util.helper.getMFN(table, obj);
}
function getMFN(table, obj) {
  return util.helper.getMFN(table, obj);
}
function getSorP(table, obj) {
  return util.helper.getSorP(table, obj);
}
//-------- using DictManager
function setWordData(word, wordData) {
  util.languageImpl.getDictManager().setWordData(word, wordData);
}
function setAdjData(word, wordData) {
  util.languageImpl.getDictManager().setAdjData(word, wordData);
}
/**
 * @license
 * Copyright 2021 Ludan Stoecklé
 * SPDX-License-Identifier: Apache-2.0
 */
function syn_fct(items) {
  return util.synManager.synFct(items);
}
function printObj(before, obj) {
  console.log(before + ' ' + JSON.stringify(obj).substring(0, 20));
}
function logParams(params) {
  console.log('params: ' + JSON.stringify(params));
}
function valueToSorP(val) {
  return util.languageImpl.isPlural(val) ? 'P' : 'S';
}
function debug(filename, line) {
  if (filename !== null) {
    pug_debug_filename = filename;
  }
  pug_debug_line = line;
  if (locals.renderDebug) {
    // locals is available
    let id = line;
    if (filename !== null) {
      id = filename.replace(/"/g, '') + ': ' + id;
    }
    pug_html = pug_html + `<span class="rosaenlg-debug" id="${id}"></span>`;
  }
}
pug_mixins["foreach"] = pug_interp = function(elts, mixinFct, asm, params){ // NAME_foreach
var block = (this && this.block), attributes = (this && this.attributes) || {};
function addToParams(_toAdd) { return Object.assign({}, params, _toAdd); }
util.asmManager.foreach(elts, mixinFct, asm, params);
};
var foreach=pug_mixins["foreach"];
pug_mixins["value"] = pug_interp = function(obj, params){ // NAME_value
var block = (this && this.block), attributes = (this && this.attributes) || {};
function addToParams(_toAdd) { return Object.assign({}, params, _toAdd); }
util.valueManager.value(obj, params);
};
var value=pug_mixins["value"];
pug_mixins["syn"] = pug_interp = function(){ // NAME_syn
var block = (this && this.block), attributes = (this && this.attributes) || {};
var items = [];
for (pug_interp = 0; pug_interp < arguments.length; pug_interp++) {
  items.push(arguments[pug_interp]);
}
function addToParams(_toAdd) { return Object.assign({}, params, _toAdd); }
util.synManager.simpleSyn(items);
};
var syn=pug_mixins["syn"];
pug_mixins["agreeAdj"] = pug_interp = function(adjective, subject, params){ // NAME_agreeAdj
var block = (this && this.block), attributes = (this && this.attributes) || {};
function addToParams(_toAdd) { return Object.assign({}, params, _toAdd); }
util.adjectiveManager.agreeAdj(adjective, subject, params);
};
var agreeAdj=pug_mixins["agreeAdj"];
pug_mixins["recipientPossession"] = pug_interp = function(owned){ // NAME_recipientPossession
var block = (this && this.block), attributes = (this && this.attributes) || {};
function addToParams(_toAdd) { return Object.assign({}, params, _toAdd); }
util.possessiveManager.recipientPossession(owned);
};
var recipientPossession=pug_mixins["recipientPossession"];
pug_mixins["thirdPossession"] = pug_interp = function(owner, owned, params){ // NAME_thirdPossession
var block = (this && this.block), attributes = (this && this.attributes) || {};
function addToParams(_toAdd) { return Object.assign({}, params, _toAdd); }
util.possessiveManager.thirdPossession(owner, owned, params);
};
var thirdPossession=pug_mixins["thirdPossession"];
pug_mixins["dummy"] = pug_interp = function(){ // NAME_dummy
var block = (this && this.block), attributes = (this && this.attributes) || {};
function addToParams(_toAdd) { return Object.assign({}, params, _toAdd); }
pug_mixins['dummy']();
};
var dummy=pug_mixins["dummy"];
pug_mixins["flushBuffer"] = pug_interp = function(){ // NAME_flushBuffer
var block = (this && this.block), attributes = (this && this.attributes) || {};
function addToParams(_toAdd) { return Object.assign({}, params, _toAdd); }
pug_html = pug_html + "¤¤";
};
var flushBuffer=pug_mixins["flushBuffer"];
pug_mixins["eatSpace"] = pug_interp = function(){ // NAME_eatSpace
var block = (this && this.block), attributes = (this && this.attributes) || {};
function addToParams(_toAdd) { return Object.assign({}, params, _toAdd); }
pug_html = pug_html + "¤EATSPACE¤";
};
var eatSpace=pug_mixins["eatSpace"];
pug_mixins["uppercaseWords"] = pug_interp = function(str){ // NAME_uppercaseWords
var block = (this && this.block), attributes = (this && this.attributes) || {};
function addToParams(_toAdd) { return Object.assign({}, params, _toAdd); }
pug_html = pug_html + "¤" + (pug_escape(null == (pug_interp = util.helper.getUppercaseWords(str)) ? "" : pug_interp)) + "¤";
};
var uppercaseWords=pug_mixins["uppercaseWords"];
pug_mixins["toHtmlList"] = pug_interp = function(obj){ // NAME_toHtmlList
var block = (this && this.block), attributes = (this && this.attributes) || {};
function addToParams(_toAdd) { return Object.assign({}, params, _toAdd); }
pug_html = pug_html + "\u003Cul\u003E";
// iterate obj
;(function(){
  var $$obj = obj;
  if ('number' == typeof $$obj.length) {
      for (var index = 0, $$l = $$obj.length; index < $$l; index++) {
        var val = $$obj[index];
if (val!=null && typeof(val) == "object") {
pug_html = pug_html + "\u003Cli\u003E" + "¤" + (pug_escape(null == (pug_interp = index) ? "" : pug_interp)) + "¤" + "¤: ¤";
pug_mixins["toHtmlList"](val);
pug_html = pug_html + "¤¤\u003C\u002Fli\u003E";
}
else {
pug_html = pug_html + "\u003Cli\u003E" + "¤" + (pug_escape(null == (pug_interp = index + ': ' + val) ? "" : pug_interp)) + "¤" + "\u003C\u002Fli\u003E";
}
      }
  } else {
    var $$l = 0;
    for (var index in $$obj) {
      $$l++;
      var val = $$obj[index];
if (val!=null && typeof(val) == "object") {
pug_html = pug_html + "\u003Cli\u003E" + "¤" + (pug_escape(null == (pug_interp = index) ? "" : pug_interp)) + "¤" + "¤: ¤";
pug_mixins["toHtmlList"](val);
pug_html = pug_html + "¤¤\u003C\u002Fli\u003E";
}
else {
pug_html = pug_html + "\u003Cli\u003E" + "¤" + (pug_escape(null == (pug_interp = index + ': ' + val) ? "" : pug_interp)) + "¤" + "\u003C\u002Fli\u003E";
}
    }
  }
}).call(this);
pug_html = pug_html + "\u003C\u002Ful\u003E";
};
var toHtmlList=pug_mixins["toHtmlList"];
pug_mixins["verb"] = pug_interp = function(subject, verbInfo, params){ // NAME_verb
var block = (this && this.block), attributes = (this && this.attributes) || {};
function addToParams(_toAdd) { return Object.assign({}, params, _toAdd); }
util.sentenceManager.verb(subject, verbInfo, params);
};
var verb=pug_mixins["verb"];
pug_mixins["subjectVerb"] = pug_interp = function(subject, verbInfo, params){ // NAME_subjectVerb
var block = (this && this.block), attributes = (this && this.attributes) || {};
function addToParams(_toAdd) { return Object.assign({}, params, _toAdd); }
util.sentenceManager.subjectVerb(subject, verbInfo, params);
};
var subjectVerb=pug_mixins["subjectVerb"];
pug_mixins["subjectVerbAdj"] = pug_interp = function(subject, verbInfo, adjective, params){ // NAME_subjectVerbAdj
var block = (this && this.block), attributes = (this && this.attributes) || {};
function addToParams(_toAdd) { return Object.assign({}, params, _toAdd); }
util.sentenceManager.subjectVerbAdj(subject, verbInfo, adjective, params);
};
var subjectVerbAdj=pug_mixins["subjectVerbAdj"];
pug_mixins["verbPart"] = pug_interp = function(){ // NAME_verbPart
var block = (this && this.block), attributes = (this && this.attributes) || {};
function addToParams(_toAdd) { return Object.assign({}, params, _toAdd); }
util.verbsManager.popVerbPartInBuffer()
};
var verbPart=pug_mixins["verbPart"];
// default value if not provided in the parameters
data = data || {
  salesQty: 1000,
  salesAmount: 500000,
}

pug_mixins['synHelper1'] = pug_interp = function(pos){
  switch(pos){
case 1:
pug_html = pug_html + "¤¤";
pug_mixins["value"](data.salesQty);
pug_html = pug_html + "¤ items were sold¤¤\n¤¤for a total amount of ¤";
pug_mixins["value"](data.salesAmount, {'FORMAT': '0a$'});
pug_html = pug_html + "¤¤";
  break;
case 2:
pug_html = pug_html + "¤total sales were ¤";
pug_mixins["value"](data.salesAmount, {'FORMAT': '0a$'});
pug_html = pug_html + "¤ for ¤";
pug_mixins["value"](data.salesQty);
pug_html = pug_html + "¤ items¤";
  break;
  }
};
var synHelper1=pug_mixins["synHelper1"];
util.synManager.runSynz(synHelper1, 'synHelper1', 2, Object.assign({}, undefined, {}));
pug_html = pug_html + "¤.¤";
    }.call(this, "JSON" in locals_for_with ?
        locals_for_with.JSON :
        typeof JSON !== 'undefined' ? JSON : undefined, "Object" in locals_for_with ?
        locals_for_with.Object :
        typeof Object !== 'undefined' ? Object : undefined, "console" in locals_for_with ?
        locals_for_with.console :
        typeof console !== 'undefined' ? console : undefined, "data" in locals_for_with ?
        locals_for_with.data :
        typeof data !== 'undefined' ? data : undefined, "embeddedLinguisticResources" in locals_for_with ?
        locals_for_with.embeddedLinguisticResources :
        typeof embeddedLinguisticResources !== 'undefined' ? embeddedLinguisticResources : undefined, "params" in locals_for_with ?
        locals_for_with.params :
        typeof params !== 'undefined' ? params : undefined, "synHelper1" in locals_for_with ?
        locals_for_with.synHelper1 :
        typeof synHelper1 !== 'undefined' ? synHelper1 : undefined, "util" in locals_for_with ?
        locals_for_with.util :
        typeof util !== 'undefined' ? util : undefined));
    ;;return locals.util.getFiltered();}
export default template;